### ISSUE TITLE
Update workflow to support a new custom exception report

### ISSUE THREAD

**Issue by daissatou2** [OpenFn/akf#13]
## Background, context, and business value
Update the AKF workflow to introduce some additional data cleaning and support a custom exception report at the end of the workflow. Regardless of whether the Salesforce upsert step succeeds or not, separately (in parallel) we want to see a report of all the data issues in the CSVs.  

## The specific request, in as few words as possible
The new workflow should have 5 jobs. Please create the two new jobs.

1. [Get Actuals from Sharepoint](https://github.com/OpenFn/akf/blob/staging/jobs/getActuals.js) (EXISTING)
2. Check If Codes Exist (**NEW**)
3. Log exceptions (**NEW**)
4. [Map & Clean Data](https://github.com/OpenFn/akf/blob/staging/jobs/applyMappings.js) (EXISTING - will have same trigger as job 3)
5. [Load Financials to AIMS ](https://openfn.org/projects/akf-bc-aims-integration/jobs/J0Cd6j)(EXISTING)

### "checkIfCodesExist" Job Logic

The BC export could contain rows with`Grant Code` `Cost Center` or `Project Code` which are blank OR which are not found in Salesforce.  The new `checkIfCodesExist` job should :
1. Identify rows from the BC export with blank `Grant Code` ,`Cost Center`, and/or `Project Code` (sometimes 1, 2, or all 3 of these will be missing)
2. Identify rows from the BC export with a `Grant Code` `Cost Center` and/or `Project Code` that is not found in Salesforce (sometimes 1, 2, or all 3 of these will not match what is in Salesforce)

**BLANK/MISSING DATA:**
If there are CSV rows in the export with a blank for any of the three columns, then do the following actions for each type of error...

1. If a row has a blank `Grant Code`
- In Job 3, we want to log these rows as exceptions: `rowsWithBlankGrantCode`
- In Job 4, we want to discard it from the data to be upserted to Salesforce --> filter from final `state.data` 

2. If a row has a blank `Cost Center`
- In Job 3, we want to log these rows as exceptions: `rowsWithBlankCostCenter`
- In Job 4, replace the blank `Cost Center` with the default value `GEN001`. Continue to upsert this data in Salesforce. 

3. If a row has a blank `Project Code`
- In Job 3, we want to log these rows as exceptions: `rowsWithBlankProjectCode`
- In Job 4, replace the blank `Project Code` with the default value `XX_UNK`. Continue to upsert this data in Salesforce. 

**MATCHING DATA NOT FOUND IN SALESFORCE:**
Run Salesforce queries to check _in bulk_  if the `Grant Code`m `Cost Center`, and/or `Project Code` in the BC export are in Salesforce. 

`SELECT Grant_Code__c FROM ampi__Project__c WHERE Grant_Code__c in (array of csv grant_codes)`
`SELECT BC_Cost_Center__c FROM Account WHERE BC_Cost_Center__c in (array of csv cost_centers)`
`SELECT Name FROM Project_Code__c WHERE Name in (array of csv project_codes)`

For each of these queries --> return a list of `matchingCodes` and `noMatchingCode` for `Grant Code`, `Cost Center`, and `Project Code` (e.g., `matchingGrantCode`, `noMatchingGrantCode`)

After keeping track of these codes in separate lists, do the following with the original BC data: 
1. If a row has no matching `Grant Code`
- In Job 3, we want to log these rows as exceptions: `noMatchingGrantCode`
- In Job 4, we want to discard it from the data to be upserted to Salesforce --> filter from final `state.data` 

2. If a row has no matching `Cost Center`
- In Job 3, we want to log these rows as exceptions: `noMatchingCostCenter`
- In Job 4, replace the blank `Cost Center` with the default value `GEN001`. Continue to upsert this data in Salesforce. 

3. If a row has no matching `Project Code`
- In Job 3, we want to log these rows as exceptions: `noMatchingProjectCode`
- In Job 4, replace the blank `Project Code` with the default value `XX_UNK`. Continue to upsert this data in Salesforce. 

**A successful run of "Check If Codes Exist" (job 2) will trigger both the "Map & Clean Data" (job4) and "Log exceptions" jobs (job 3).** 
1. The initial state of "Map & Clean Data" should be the list of filtered rows to import to Salesforce (following the logic of removing and updating rows above). 
2. The initial state of "Log exceptions" should be the different arrays that contain bad rows and bad codes that we will want to log for the client to review later: `rowsWithBlankGrantCode`, `rowsWithBlankCostCenter`, `rowsWithBlankProjectCode` `noMatchingGrantCode`, `noMatchingCostCenter`, and `noMatchingProjectCode`. 

### "logExceptions" Job Logic
This job should output the content of the arrays that contain the problem data. Please follow the format in this this [sample logging report](https://docs.google.com/document/d/1LbXlo71r2oDW5dBjXy3Ftu26QzCRzu_PwqBOUkVVDdQ/edit). Reminder: for blank/missing data, we should output the entire row, for non-matching data, we should output just the code for which a match has not been found (i.e. only the grant code, project code, or cost center). This job should always pass. 

### "applyMappings" job update
The only update to this job is the initial state. Instead of all of the BC rows, the initial state will be only the "cleaned rows"

### Links and diagrams
-[Mapping specifications](https://docs.google.com/spreadsheets/d/1h6Vnx7jR_PHbSiv1H33O51580lLMEPAvodPqJqbD9MA/edit#gid=1275153608)

![image](https://github.com/OpenFn/akf/assets/80456839/af3b34cc-5872-497f-bf19-3ecb10b222e8)

### Expected data volumes
It's looking like some of these CSV files might be up to 15 MB if they contain 12 months of data.

## state.json

Job 1: Get Actuals from Sharepoint 
```json
{
  "configuration": { "See LP: AKF Sharepoint Site" },
  "data": { "NA"},
}
```

Job 2: Check If Codes Exist 

```json
{
  "configuration": { NA },
  "data": { "generated by job 1"},
}
```

Job 3: Map & Clean Data

```json
{
  "configuration": { NA },
  "data": { "generated by job 2"},
}
```

Job 4: Load Financials to AIMS
```json
{
  "configuration": { "See LP: AKF Salesforce Sandbox - Partial 2023" },
  "data": { "generated by job 3"},
}
```

Job 5:  Log exceptions
```json
{
  "configuration": { "NA" },
  "data": { "generated by job 2"},
}
```

## expression.js

NEW JOBS:
1. Check If Codes Exists: http, flow job (parent Get Actuals from Sharepoint), output - cleanedRows, badRows
2. Log Exceptions: http, flow job (parent Check If Codes Exists), output - exception report 

## Toggl 
AKF BC M4 - Testing & Iteration

**Comment by daissatou2** [comment]
@mtuchi when this is ready let's first test with the file named `bc_actuals_2023_09_11.csv` by using that date as the cursor. The counts and log format should match what is in this sample [report](https://docs.google.com/document/d/1LbXlo71r2oDW5dBjXy3Ftu26QzCRzu_PwqBOUkVVDdQ/edit).

**Comment by aleksa-krolls** [comment]
@mtuchi related to our conversation yesterday, this is ready to be worked on. See Aicha's comment above re: what file to use. I will be available to chat on the phone after 10:30 EAT, in case you have any questions.

**Comment by daissatou2** [comment]
@mtuchi for this query: `SELECT Grant_Code__c FROM ampi__Project__c WHERE Grant_Code__c in (array of csv grant_codes)` please make sure that the trailing "M" and "E" are removed from the list of grant codes before querying. See #6 for background on these trailing characters. 

For example, if we have a list of Grant Codes from the BC export like this: 
111GC3
GAF104 
GAF104E
JXN01M

Then we want our query to look like...
SELECT Grant_Code__c FROM ampi__Project__c WHERE Grant_Code__c in (111GC3, GAF104, JXN01). 


Notice how the trailing "M" and "E" have been removed from the last two grant codes before querying.

**Comment by aleksa-krolls** [comment]
@mtuchi as discussed let's try (1) first querying from Salesforce to get list of existing codes, and (2) then compare with what's in the CSV. 
```
SELECT Id, Name from Project_Code__c
```
```
SELECT BC_Cost_Center__c 
```
Right now, in sandbox these queries will _not_ return more than ~1800 records, but in the future... they might return more than 2000 records. Therefore, we should write the queries to page through the results if there are >2000 records (to get around the Salesforce query limit of 2000 records per request). 

To simulate paging to build^ this out, you can add `LIMIT 100` to this query (which typically will return 1795 records). 
```
SELECT Id, Name from Project_Code__c LIMIT 100
```

@daissatou2 the client's original [list of BC Cost Centers](https://docs.google.com/spreadsheets/d/1WshRpUl92UiQGA_2aRsyhiQM1I5IUKmIWVC2E9LYS-U/edit#gid=810609412) only contained 237 records... but Mtuchi is detecting over 7000 in the source data ❗This is WAY more than we originally anticipated. Let's ask the client what is going on here.

**Comment by mtuchi** [comment]
I have added the changes suggested above in this PR #14, I am going to merge on `staging` so that i can test the limits of data passed between jobs 
cc @daissatou2 @aleksa-krolls

**Comment by daissatou2** [comment]
I tested with both the "bc_actuals_2023_8_11" file and the "bc_actuals_2023_09_12" file . I have left my feedback on the 8/11 file in [Slack](https://openfn.slack.com/archives/C017ELVRSM8/p1694711278287549?thread_ts=1694702507.511129&cid=C017ELVRSM8). 

**bc_actuals_2023_09_12 file feedback:**

**Log Exceptions Job**, see [log](https://www.openfn.org/projects/akf-bc-aims-integration/runs/06503714-6497-7012-b800-3d052dadf632): 
@mtuchi the numbers seem to be off for most of these...please double check. The expected output for this file along with counts are [here](https://docs.google.com/spreadsheets/d/1IlerWIAA1BUO3upQ5TLSMR6wTsWHCVbvchUg5LsFv44/edit?pli=1#gid=904971764). Other issues beside the counts: 


1. Line 5 of the log says 2 grant codes were not found but only 1 is in the array. The other item in the array is an empty string. We should deduplicate the list of grant codes and the count in the description should match the number of items in the array. This is also the case for the cost center and project code arrays.
2. There's something up with the "Sum of Actuals" being shown in this log. For example "197624.5" on line 71 is not the CSV...this looks like it might be an aggregated sum. These arrays should reflect what is in the BC data so that the client can go find the row in their original file for troubleshooting.
3. Not as high priority since they might go with CSV output, but is it easy enough to output the row values with the keys? Otherwise might be hard to read. 


Load financials to AIMS Job, see [log](https://www.openfn.org/projects/akf-bc-aims-integration/runs/06503714-70c4-7333-9e39-a97cc430bd7a):
1. Why are we upserting record with id “1EC10_default_XX_UNK_AFHS01_1_2022_Afghanistan AKF” three times? Looks like this was not aggregated for some reason...see table 4 in the excepted out sheet for the aggregation of records that should be loaded to SF. 
2. Not sure if this is an issue but is there are reason why some of the "Amount_Actual_Local_Currency__c" have so many decimal places and others only have 2? Line 15 for example. The BC data only has two decimal places.

**Comment by daissatou2** [comment]
@mtuchi I know you're focusing on the adaptor today but wanted to give you a heads-up that we will be moving forward with the CSV report. That means everything related to the logic for this issue will remain the same but the final output will be a CSV file instead of an OpenFn log. I will create a separate issue for generating the CSV.

**Comment by daissatou2** [comment]
Something to note: 

The original requirement in this issue was if codes were not found in Salesforce (`noMatchingGrantCode`, `noMatchingCostCenter`, and `noMatchingProjectCode`), to log the code and not the entire row. With the new approach of generating an exception report XLS, **the entire row should posted to the file just like what we are doing for the rows with blank codes.**

See related exception report file issue #15.

**Comment by mtuchi** [comment]
@daissatou2 i have made changes to the workflow to handle exceptions report in the PR #17, 
I am going to merge for testing on platform app
cc @aleksa-krolls

**Comment by daissatou2** [comment]
Closing this out. The exception report is being tracked in issue #15 .
